File: class09/python/GQS0901-CoberturaDecisao.py
#!/usr/bin/env python3
"""
GQS0901 - Demonstração de Cobertura de Instrução e Decisão.
Analisa um código exemplo e calcula as métricas de cobertura.
Uso: python3 GQS0901-CoberturaDecisao.py
"""


def funcao_exemplo(a: int, b: int) -> str:
    """Função exemplo para análise de cobertura."""
    if a > 0 and b > 0:
        return "Ambos positivos"
    elif a > 0:
        return "A positivo"
    elif b > 0:
        return "B positivo"
    else:
        return "Nenhum positivo"


def calcular_complexidade_ciclomatica(decisions: int) -> int:
    """M = número de decisões + 1 (McCabe)."""
    return decisions + 1


def main():
    print("=" * 60)
    print("  GQS0901 - COBERTURA DE INSTRUÇÃO E DECISÃO")
    print("=" * 60)

    print("\nFunção analisada: funcao_exemplo(a, b)")
    print("  if a > 0 and b > 0 → Ambos positivos")
    print("  elif a > 0 → A positivo")
    print("  elif b > 0 → B positivo")
    print("  else → Nenhum positivo")
    print(f"\n  Complexidade Ciclomática (M) = {calcular_complexidade_ciclomatica(3)}")
    print("  (3 decisões + 1 = 4 caminhos independentes)")

    print("\n--- Cobertura de Decisão ---")
    casos = [
        (5, 5, "Ambos positivos"),
        (5, -1, "A positivo"),
        (-1, 5, "B positivo"),
        (-1, -1, "Nenhum positivo"),
    ]
    print(f"{'Teste (a, b)':<15} {'Resultado':<20} {'Cobertura':<15}")
    print("-" * 50)
    for i, (a, b, esperado) in enumerate(casos, 1):
        resultado = funcao_exemplo(a, b)
        coberto = "✅" if resultado == esperado else "❌"
        print(f"  CT-{i:02d} ({a:<2},{b:<3}) {resultado:<20} {coberto}")

    print("\n  Para 100% de cobertura de decisão: todos os 4 casos")
    print("  Para 100% de cobertura de instrução: casos 1, 2, 3, 4")


if __name__ == "__main__":
    main()

File: class09/java/GQS0901-CoberturaDecisao.java
/**
 * GQS0901 - Demonstração de Cobertura de Instrução e Decisão.
 * Analisa uma função exemplo e calcula as métricas de cobertura e a
 * complexidade ciclomática (McCabe).
 *
 * Compilação e execução:
 *   javac GQS0901-CoberturaDecisao.java
 *   java -cp . GQS0901_CoberturaDecisao
 */
import java.util.*;

/** Representa um caso de teste com entradas e o resultado esperado. */
class CasoTeste {
    int a;
    int b;
    String esperado;

    CasoTeste(int a, int b, String esperado) {
        this.a = a;
        this.b = b;
        this.esperado = esperado;
    }
}

class GQS0901_CoberturaDecisao {

    /** Função exemplo para análise de cobertura. */
    static String funcaoExemplo(int a, int b) {
        if (a > 0 && b > 0) {
            return "Ambos positivos";
        } else if (a > 0) {
            return "A positivo";
        } else if (b > 0) {
            return "B positivo";
        } else {
            return "Nenhum positivo";
        }
    }

    /** M = número de decisões + 1 (McCabe). */
    static int calcularComplexidadeCiclomatica(int decisoes) {
        return decisoes + 1;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("  GQS0901 - COBERTURA DE INSTRUÇÃO E DECISÃO");
        System.out.println("=".repeat(60));

        System.out.println("\nFunção analisada: funcaoExemplo(a, b)");
        System.out.println("  if a > 0 e b > 0 → Ambos positivos");
        System.out.println("  else if a > 0 → A positivo");
        System.out.println("  else if b > 0 → B positivo");
        System.out.println("  else → Nenhum positivo");
        System.out.printf("%n  Complexidade Ciclomática (M) = %d%n", calcularComplexidadeCiclomatica(3));
        System.out.println("  (3 decisões + 1 = 4 caminhos independentes)");

        System.out.println("\n--- Cobertura de Decisão ---");
        List<CasoTeste> casos = List.of(
            new CasoTeste(5, 5, "Ambos positivos"),
            new CasoTeste(5, -1, "A positivo"),
            new CasoTeste(-1, 5, "B positivo"),
            new CasoTeste(-1, -1, "Nenhum positivo")
        );
        System.out.printf("%-15s %-20s %-15s%n", "Teste (a, b)", "Resultado", "Cobertura");
        System.out.println("-".repeat(50));
        for (int i = 0; i < casos.size(); i++) {
            CasoTeste ct = casos.get(i);
            String resultado = funcaoExemplo(ct.a, ct.b);
            String coberto = resultado.equals(ct.esperado) ? "✅" : "❌";
            System.out.printf("  CT-%02d (%-2d,%-3d) %-20s %s%n", i + 1, ct.a, ct.b, resultado, coberto);
        }

        System.out.println("\n  Para 100% de cobertura de decisão: todos os 4 casos");
        System.out.println("  Para 100% de cobertura de instrução: casos 1, 2, 3, 4");
    }
}
